providers/coordinator.go

```diff
@@ -426,48 +236,199 @@ func (c *coordinator) unsafeSetAssetRuntime(asset *inventory.Asset, runtime *Run
 	return found
 }
 
-func (c *coordinator) Stop(provider *RunningProvider) error {
-	if provider == nil {
-		return nil
-	}
+// RemoveRuntime will remove a runtime from the coordinator. This can potentially
+// shutdown a running provider if it's not referenced by another runtime.
+func (c *coordinator) RemoveRuntime(runtime *Runtime) {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
+	c.unsafeRefreshRuntimes()
 
-	found := c.RunningByID[provider.ID]
-	if found != nil {
-		delete(c.RunningByID, provider.ID)
+	if runtime.asset() != nil {
+		delete(c.runtimes, runtime.asset().Mrn)
+		for _, id := range runtime.asset().PlatformIds {
+			delete(c.runtimes, id)
+		}
+	}
+
+	// Analyze the providers that are still being used by active runtimes
+	usedProviders := map[string]struct{}{}
+	for _, r := range c.runtimes {
+		for _, p := range r.providers {
+			usedProviders[p.Instance.ID] = struct{}{}
+		}
+	}
+
+	// Shutdown any providers that are not being used anymore
+	for id, p := range c.runningByID {
+		if _, ok := usedProviders[id]; !ok {
+			log.Debug().Msg("shutting down unused provider " + p.Name)
+			if err := c.stop(p); err != nil {
+				log.Warn().Err(err).Str("provider", p.Name).Msg("failed to shut down provider")
+			}
+		}
+	}
+}
+
+func (c *coordinator) GetRunningProvider(id string, update UpdateProvidersConfig) (*RunningProvider, error) {
```

Comment:
Is it possible for 2 almost synchronous calls to be made to this function that cause the provider to be started twice? Asking because running may be nil on the request and startProvider is not in any sync lock, i.e. I wonder if two functions of startProvider can run before `runningByID` gets set

Reply:
nice catch! i made sure we have a lock for the whole duration of `GetRunningProvider`. That should make sure we don't run into such situations